File: src/topos/cli/commands/coverage.py
from __future__ import annotations

import json
import sys
from dataclasses import asdict
from pathlib import Path

import click

from topos.graphs.ast.dispatch import SUPPORTED_LANGUAGES, parse_source

_EVALUATE_LANGUAGE_CHOICE = click.Choice(sorted(SUPPORTED_LANGUAGES))


@click.command("structural-test-coverage")
@click.option(
    "--language",
    default="python",
    show_default=True,
    type=_EVALUATE_LANGUAGE_CHOICE,
    help="Language for tree-sitter / UAST parsing of all listed files.",
)
@click.option(
    "--tests",
    "test_paths",
    type=click.Path(exists=True, dir_okay=False),
    multiple=True,
    required=True,
    help="Test file path (repeat for multiple test modules).",
)
@click.option(
    "--k",
    "kgram_length",
    default=3,
    show_default=True,
    type=int,
    help="Length of each DFS kind n-gram for path recall.",
)
@click.option(
    "--include-unknown",
    is_flag=True,
    help="Count Unknown UAST kinds in histograms and k-grams.",
)
@click.option(
    "--json",
    "output_json_flag",
    is_flag=True,
    help="Emit a single JSON object with scores and diagnostics.",
)
@click.option(
    "--coverage-threshold",
    default=0.5,
    show_default=True,
    type=float,
    help="Minimum best-match recall to count a PUT declaration as covered.",
)
@click.argument(
    "put_paths",
    nargs=-1,
    type=click.Path(exists=True, dir_okay=False),
    required=True,
)
def structural_test_coverage_cmd(
    put_paths: tuple[str, ...],
    test_paths: tuple[str, ...],
    language: str,
    kgram_length: int,
    include_unknown: bool,
    output_json_flag: bool,
    coverage_threshold: float,
) -> None:
    """Structural overlap of tests toward the program-under-test (UAST)."""
    if kgram_length < 1:
        click.echo("Error: --k must be >= 1.", err=True)
        sys.exit(1)

    put_roots: list[object] = []
    for path in put_paths:
        source = Path(path).read_text(encoding="utf-8")
        result = parse_source(source=source, language=language, file=str(path))
        if result.uast_root is None:
            click.echo(f"Error: No UAST root for PUT file {path}.", err=True)
            sys.exit(1)
        put_roots.append(result.uast_root)

    test_roots: list[object] = []
    for path in test_paths:
        source = Path(path).read_text(encoding="utf-8")
        result = parse_source(source=source, language=language, file=str(path))
        if result.uast_root is None:
            click.echo(f"Error: No UAST root for test file {path}.", err=True)
            sys.exit(1)
        test_roots.append(result.uast_root)

    from topos.evaluation.policies.coverage import score_declaration_coverage
    from topos.functors.profunctors.uast.structural_test_coverage import (
        declaration_coverage,
    )

    report = declaration_coverage(
        put_roots,
        test_roots,
        k=kgram_length,
        include_unknown=include_unknown,
    )
    decision = score_declaration_coverage(report, threshold=coverage_threshold)

    if output_json_flag:
        payload = asdict(report)
        payload.update(asdict(decision))
        payload["language"] = language
        payload["put_paths"] = list(put_paths)
        payload["test_paths"] = list(test_paths)
        click.echo(json.dumps(payload, indent=2))
        return

    click.echo("Structural test coverage (declaration-level bipartite)")
    click.echo(f"Language: {language}")
    click.echo(f"PUT files ({len(put_paths)}): {', '.join(put_paths)}")
    click.echo(f"Test files ({len(test_paths)}): {', '.join(test_paths)}")
    click.echo()
    click.echo("Declaration Coverage")
    click.echo("-" * 52)
    click.echo(f"  Mean declaration coverage:  {report.mean_declaration_coverage:.4f}")
    click.echo(f"  Declaration coverage rate:  {decision.coverage_rate:.4f}")
    click.echo(f"  Coverage threshold:         {decision.threshold:.2f}")
    click.echo(f"  PUT declarations:           {report.put_declaration_count}")
    click.echo(f"  Test declarations:          {report.test_declaration_count}")
    click.echo()
    click.echo("Category-Stratified Recall (disjoint)")
    click.echo("-" * 52)
    click.echo(f"  Statement recall:           {report.stmt_recall:.4f}")
    click.echo(f"  Expression recall:          {report.expr_recall:.4f}")
    click.echo()
    click.echo("Precision and F-score")
    click.echo("-" * 52)
    click.echo(f"  Mean test precision:        {report.mean_test_precision:.4f}")
    click.echo(f"  F2 score (beta=2):          {decision.f2_score:.4f}")
    click.echo()
    click.echo(f"Path Recall (declaration-scoped k={report.k} grams)")
    click.echo("-" * 52)
    click.echo(f"  Decl path recall:           {report.declaration_path_recall_kgram:.4f}")
    
    uncovered = decision.uncovered_declarations
    if uncovered:
        click.echo()
        click.echo(f"Uncovered PUT declarations (below {decision.threshold:.0%})")
        click.echo("-" * 52)
        for loc, score in uncovered:
            click.echo(f"  {loc}  (best score: {score:.3f})")

File: crates/topos-cli/src/commands/coverage.rs
//! `topos coverage` — structural (UAST) test coverage between a
//! program-under-test and its test suite.
//!
//! Ported from `topos/cli/commands/coverage.py`, calling straight into
//! [`structural_test_coverage::declaration_coverage`] and
//! [`policies::coverage::score_declaration_coverage`]. The Python
//! original's JSON output mode and MCP-schema-shaped payload are not
//! ported here — this pass is plain-text only (see issue #147 scope).

use std::path::PathBuf;

use clap::Args;
use topos_core::evaluation::policies::coverage::score_declaration_coverage;
use topos_core::functors::profunctors::uast::structural_test_coverage::declaration_coverage;
use topos_core::graphs::ast::dispatch::parse_source;
use topos_core::graphs::ast::languages::SUPPORTED_LANGUAGES;
use topos_core::graphs::uast::models::UASTNode;

#[derive(Args)]
pub struct CoverageArgs {
    /// Program-under-test file(s).
    #[arg(required = true)]
    pub put_paths: Vec<PathBuf>,
    /// Test file path (repeat for multiple test modules).
    #[arg(long = "tests", required = true)]
    pub test_paths: Vec<PathBuf>,
    /// Language for tree-sitter / UAST parsing of all listed files.
    #[arg(long, default_value = "python")]
    pub language: String,
    /// Length of each DFS kind n-gram for path recall.
    #[arg(long = "k", default_value_t = 3)]
    pub kgram_length: usize,
    /// Count Unknown UAST kinds in histograms and k-grams.
    #[arg(long)]
    pub include_unknown: bool,
    /// Minimum threshold for coverage policies to pass.
    #[arg(long, default_value_t = 0.5)]
    pub coverage_threshold: f64,
}

fn parse_uast_roots(paths: &[PathBuf], language: &str) -> Result<Vec<UASTNode>, String> {
    paths
        .iter()
        .map(|path| {
            let source = std::fs::read_to_string(path)
                .map_err(|e| format!("reading {}: {e}", path.display()))?;
            let file = path.to_string_lossy().into_owned();
            let result = parse_source(&source, language, Some(&file))
                .map_err(|e| format!("parsing {}: {e}", path.display()))?;
            Ok(result.uast_root)
        })
        .collect()
}

pub fn run(args: CoverageArgs) -> Result<(), String> {
    if args.kgram_length < 1 {
        return Err("--k must be >= 1".to_string());
    }
    if !SUPPORTED_LANGUAGES.contains(&args.language.as_str()) {
        return Err(format!(
            "unsupported language '{}' (expected one of: {})",
            args.language,
            SUPPORTED_LANGUAGES.join(", ")
        ));
    }

    let put_roots = parse_uast_roots(&args.put_paths, &args.language)?;
    let test_roots = parse_uast_roots(&args.test_paths, &args.language)?;
    let put_refs: Vec<&UASTNode> = put_roots.iter().collect();
    let test_refs: Vec<&UASTNode> = test_roots.iter().collect();

    let report = declaration_coverage(
        &put_refs,
        &test_refs,
        args.kgram_length,
        args.include_unknown,
    )
    .map_err(|e| e.to_string())?;
    let decision = score_declaration_coverage(&report, args.coverage_threshold);

    println!("Topos Structural Test Coverage");
    println!("Language: {}", args.language);
    println!(
        "PUT files ({}): {}",
        args.put_paths.len(),
        display_paths(&args.put_paths)
    );
    println!(
        "Test files ({}): {}",
        args.test_paths.len(),
        display_paths(&args.test_paths)
    );
    println!();

    println!("UAST Declaration-Level Coverage");
    println!("{}", "-".repeat(52));
    println!(
        "  Mean declaration coverage:  {:.4}",
        report.mean_declaration_coverage
    );
    println!(
        "  Declaration coverage rate:  {:.4}",
        decision.coverage_rate
    );
    println!("  Coverage threshold:         {:.2}", decision.threshold);
    println!(
        "  PUT declarations:           {}",
        report.put_declaration_count
    );
    println!(
        "  Test declarations:          {}",
        report.test_declaration_count
    );
    println!();

    println!("Category-Stratified Recall (disjoint)");
    println!("{}", "-".repeat(52));
    println!("  Statement recall:           {:.4}", report.stmt_recall);
    println!("  Expression recall:          {:.4}", report.expr_recall);
    println!();

    println!("Precision and F-score");
    println!("{}", "-".repeat(52));
    println!(
        "  Mean test precision:        {:.4}",
        report.mean_test_precision
    );
    println!("  F2 score (beta=2):          {:.4}", decision.f2_score);
    println!();

    println!("Path Recall (declaration-scoped k={} grams)", report.k);
    println!("{}", "-".repeat(52));
    println!(
        "  Decl path recall:           {:.4}",
        report.declaration_path_recall_kgram
    );

    if !decision.uncovered_declarations.is_empty() {
        println!();
        println!(
            "Uncovered PUT declarations (below {:.0}%)",
            decision.threshold * 100.0
        );
        println!("{}", "-".repeat(52));
        let mut uncovered = decision.uncovered_declarations.clone();
        uncovered.sort_by(|a, b| a.1.total_cmp(&b.1));
        for (loc, score) in &uncovered {
            println!("  {loc}  (best score: {score:.3})");
        }
    } else {
        println!();
        println!("All measured declarations meet the threshold.");
    }

    Ok(())
}

fn display_paths(paths: &[PathBuf]) -> String {
    paths
        .iter()
        .map(|p| p.display().to_string())
        .collect::<Vec<_>>()
        .join(", ")
}
